webots_ros2_driver/src/plugins/static/Ros2Compass.cpp

```diff
@@ -0,0 +1,110 @@
+    Ros2SensorPlugin::init(node, parameters);
+    // This parameter is read when loading the URDF file
+    mDeviceName = parameters.count("name") ? parameters["name"] : "compass";
+
+    mCompass = wb_robot_get_device(mDeviceName.c_str());
+    mIsEnabled = false;
+
+    assert(mCompass != 0);
+
+    // Data publisher
+    mDataPublisher =
+      mNode->create_publisher<sensor_msgs::msg::MagneticField>(mTopicName + "/values", rclcpp::SensorDataQoS().reliable());
+    // Services
+    mEnableService = mNode->create_service<webots_ros2_msgs::srv::SetInt>(
+      mTopicName + "/enable", std::bind(&Ros2Compass::enable_callback, this, _1, _2));
+    mDisableService = mNode->create_service<webots_ros2_msgs::srv::SetInt>(
+      mTopicName + "/disable", std::bind(&Ros2Compass::disable_callback, this, _1, _2));
+    mGetSamplingPeriodService = mNode->create_service<webots_ros2_msgs::srv::GetInt>(
+      mTopicName + "/get_sampling_period", std::bind(&Ros2Compass::get_sampling_period_callback, this, _1, _2));
+    mGetLookupTableService = mNode->create_service<webots_ros2_msgs::srv::GetFloatArray>(
+      mTopicName + "/get_lookup_table", std::bind(&Ros2Compass::get_lookup_table_callback, this, _1, _2));
+    RCLCPP_DEBUG(rclcpp::get_logger(mDeviceName), (mDeviceName + " initialized!").c_str());
+
+    // Calculate timestep
+    if (mAlwaysOn) {
+      wb_compass_enable(mCompass, mPublishTimestepSyncedMs);
+      RCLCPP_DEBUG(rclcpp::get_logger(mDeviceName), (mDeviceName + " activated!").c_str());
+      mIsEnabled = true;
+    }
+  }
+  void Ros2Compass::step() {
+    if (!preStep())
+      return;
+
+    if (mIsEnabled)
+      publishData();
+
+    if (mAlwaysOn)
+      return;
+
+    mustPublish();
+  }
+  bool Ros2Compass::mustPublish() {
+    // Enable/Disable sensor
+    const bool shouldBeEnabled = mDataPublisher->get_subscription_count() > 0;
+    if (shouldBeEnabled != mIsEnabled) {
+      if (shouldBeEnabled)
+        wb_compass_enable(mCompass, mPublishTimestepSyncedMs);
+      else
+        wb_compass_disable(mCompass);
+      mIsEnabled = shouldBeEnabled;
+    }
+    return mIsEnabled;
+  }
+  void Ros2Compass::publishData() {
+    const double *compass_values = wb_compass_get_values(mCompass);
+    mDataMessage.header.stamp = mNode->get_clock()->now();
+    mDataMessage.magnetic_field.x = compass_values[0];
+    mDataMessage.magnetic_field.y = compass_values[1];
+    mDataMessage.magnetic_field.z = compass_values[2];
+    mDataPublisher->publish(mDataMessage);
+  }
+  void Ros2Compass::enable_callback(const std::shared_ptr<webots_ros2_msgs::srv::SetInt::Request> request,
+                                    std::shared_ptr<webots_ros2_msgs::srv::SetInt::Response> response) {
+    if (request->value) {
+      wb_compass_enable(mCompass, mPublishTimestepSyncedMs);
+      mAlwaysOn = true;
+      mIsEnabled = true;
+    }
+    response->success = true;
```

Comment:
After the last comments I should also change this lines right? using `request->value` as the time step argument, right?
```suggestion
   wb_compass_enable(mCompass, request->value);
   mAlwaysOn = true;
   mIsEnabled = true;
   response->success = true;
```

Reply:
Yes, you are right. I missed this one. The requested value should be used and not the global timestep.